File: scripts/download_real_assets.py
import urllib.request
import json
import os
import ssl

ctx = ssl.create_default_context()
ctx.check_hostname = False
ctx.verify_mode = ssl.CERT_NONE

headers = {
    'User-Agent': 'MaraudersGuessr/1.0 (https://github.com/marauders-guessr; contact@example.com)'
}

output_dir = os.path.join(os.path.dirname(__file__), '..', 'public', 'panoramas')
os.makedirs(output_dir, exist_ok=True)

# Wikimedia queries for real Hogwarts locations
queries = {
    'great-hall': 'Christ Church Hall Oxford interior panorama',
    'potions-dungeon': 'Lacock Abbey cloisters panorama',
    'corridors': 'Gloucester Cathedral Cloister panorama',
    'hospital-wing': 'Divinity School Oxford interior panorama',
    'library': 'Duke Humfrey Library Oxford panorama',
    'quidditch-pitch': 'Alnwick Castle courtyard panorama',
    'platform-9-3-4': 'Glenfinnan Viaduct railway steam train panorama',
    'diagon-alley': 'Leadenhall Market London interior panorama',
    'transfiguration': 'Durham Cathedral Chapter House interior panorama',
    'hogsmeade': 'Goathland railway station panorama'
}

def search_wikimedia(query):
    url = f"https://commons.wikimedia.org/w/api.php?action=query&generator=search&gsrsearch={urllib.parse.quote(query)}&gsrnamespace=6&gsrlimit=3&prop=imageinfo&iiprop=url|mime|size&format=json"
    req = urllib.request.Request(url, headers=headers)
    try:
        with urllib.request.urlopen(req, context=ctx, timeout=10) as resp:
            data = json.loads(resp.read().decode('utf-8'))
            pages = data.get('query', {}).get('pages', {})
            for page_id, page in pages.items():
                imageinfo = page.get('imageinfo', [])
                if imageinfo:
                    img_url = imageinfo[0].get('url')
                    mime = imageinfo[0].get('mime')
                    if img_url and ('image/jpeg' in mime or 'image/png' in mime or 'image/webp' in mime):
                        return img_url
    except Exception as e:
        print(f"Error searching for {query}: {e}")
    return None

results = {}
for key, query in queries.items():
    print(f"Searching for {key} ({query})...")
    img_url = search_wikimedia(query)
    if img_url:
        print(f"Found: {img_url}")
        target_file = os.path.join(output_dir, f"{key}.jpg")
        try:
            req = urllib.request.Request(img_url, headers=headers)
            with urllib.request.urlopen(req, context=ctx, timeout=20) as resp, open(target_file, 'wb') as f:
                f.write(resp.read())
            print(f"Saved {key}.jpg ({os.path.getsize(target_file)} bytes)")
            results[key] = f"/panoramas/{key}.jpg"
        except Exception as e:
            print(f"Failed to download {img_url}: {e}")
    else:
        print(f"No image found for {query}")

print("Download summary:", results)

File: scripts/download_real_assets.js
import https from 'https';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const outputDir = path.join(__dirname, '..', 'public', 'panoramas');
if (!fs.existsSync(outputDir)) {
  fs.mkdirSync(outputDir, { recursive: true });
}

const searchTerms = {
  'great-hall': 'Christ Church Hall Oxford',
  'potions-dungeon': 'Lacock Abbey Cloisters',
  'gryffindor-common-room': 'Divinity School Bodleian Library Oxford',
  'slytherin-common-room': 'Wieliczka Chapel',
  'ravenclaw-tower': 'Christ Church Cathedral Choir Oxford',
  'hufflepuff-basement': 'Gloucester Cathedral Cloisters',
  'dumbledores-office': 'Durham Cathedral interior',
  'library-restricted': 'Duke Humfrey Library Oxford',
  'astronomy-tower': 'Alnwick Castle',
  'room-of-requirement': 'Lloyds Building Interior',
  'quidditch-pitch': 'Alnwick Castle Inner Bailey',
  'hagrids-hut': 'Wistmans Wood Dartmoor',
  'herbology-greenhouses': 'Palm House Kew Gardens',
  'diagon-alley-ollivanders': 'Leadenhall Market London',
  'gringotts-vaults': 'Salina Turda mine',
  'knockturn-alley': 'The Shambles York',
  'hogsmeade-high-street': 'Goathland railway station',
  'shrieking-shack': 'Old wooden abandoned house forest',
  'ministry-atrium': 'Grand Central Terminal Inside',
  'department-of-mysteries': 'CERN LHC Tunnel',
  'platform-9-3-4': 'Glenfinnan Viaduct Jacobite',
  'forbidden-forest-clearing': 'Moss covered trees rainforest temperate'
};

function searchWiki(term) {
  const url = `https://commons.wikimedia.org/w/api.php?action=query&generator=search&gsrsearch=${encodeURIComponent(term)}&gsrnamespace=6&gsrlimit=1&prop=imageinfo&iiprop=url&iiurlwidth=2048&format=json`;
  return new Promise((resolve, reject) => {
    https.get(url, { headers: { 'User-Agent': 'MaraudersGuessr/1.0 (academic research test)' } }, (res) => {
      let data = '';
      res.on('data', chunk => data += chunk);
      res.on('end', () => {
        try {
          const json = JSON.parse(data);
          const pages = json.query?.pages || {};
          for (const [id, page] of Object.entries(pages)) {
            const thumb = page.imageinfo?.[0]?.thumburl || page.imageinfo?.[0]?.url;
            if (thumb) return resolve(thumb);
          }
          resolve(null);
        } catch (e) {
          reject(e);
        }
      });
    }).on('error', reject);
  });
}

function downloadFile(url, dest) {
  return new Promise((resolve, reject) => {
    https.get(url, { headers: { 'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36' } }, (res) => {
      if (res.statusCode >= 300 && res.statusCode < 400 && res.headers.location) {
        return downloadFile(res.headers.location, dest).then(resolve).catch(reject);
      }
      if (res.statusCode !== 200) {
        return reject(new Error(`HTTP ${res.statusCode}`));
      }
      const stream = fs.createWriteStream(dest);
      res.pipe(stream);
      stream.on('finish', () => {
        stream.close();
        resolve();
      });
    }).on('error', reject);
  });
}

function sleep(ms) {
  return new Promise(r => setTimeout(r, ms));
}

async function run() {
  console.log("Searching and downloading real photographs/panoramas for all Hogwarts chambers...");
  let idx = 0;
  for (const [key, term] of Object.entries(searchTerms)) {
    console.log(`[${++idx}/22] Searching for ${key} ("${term}")...`);
    try {
      const url = await searchWiki(term);
      if (url) {
        console.log(`  Found: ${url.slice(0, 60)}...`);
        const dest = path.join(outputDir, `${key}.jpg`);
        await downloadFile(url, dest);
        const sz = fs.statSync(dest).size;
        console.log(`  ✓ Saved ${key}.jpg (${(sz / 1024).toFixed(0)} KB)`);
      } else {
        console.warn(`  ✗ No result for ${term}`);
      }
    } catch (e) {
      console.error(`  ✗ Error: ${e.message}`);
    }
    await sleep(400);
  }
  console.log("Download process finished!");
}

run();
